snuba/utils/bucket_timer.py

```diff
@@ -0,0 +1,104 @@
+
+from snuba import environment
+from snuba.utils.metrics.wrapper import MetricsWrapper
+
+WINDOW_SIZE = timedelta(minutes=10)  # minutes
+
+
+class Bucket:
+    def __init__(
+        self,
+        project_id: Optional[int],
+        minute: datetime,
+        processing_time: timedelta = timedelta(seconds=0),
+    ) -> None:
+        self.project_id = project_id  # is None for global bucket
+        self.minute = minute
+        self.processing_time = processing_time
+
+    def add_processing_time_seconds(self, seconds: timedelta) -> None:
+        self.processing_time += seconds
+
+
+class Counter:
+    def __init__(self, consumer_group: str) -> None:
+        self.consumer_group: str = consumer_group
+        self.buckets: Deque[Bucket] = deque()
+
+    def get_time_sum(self, group: List[timedelta]) -> timedelta:
+        return sum([processing_time for processing_time in group], timedelta())
+
+    def floor_minute(self, time: datetime) -> datetime:
+        return time - timedelta(seconds=time.second, microseconds=time.microsecond)
+
+    def ceil_minute(self, time: datetime) -> datetime:
+        if time.second == 0 and time.microsecond == 0:
+            return time
+        return self.floor_minute(time + timedelta(minutes=1))
+
+    def trim_expired_buckets(self, now: datetime) -> None:
+        current_minute = self.floor_minute(now)
+        window = current_minute - WINDOW_SIZE
+        while self.buckets and self.buckets[0].minute < window:
+            self.buckets.popleft()
+
+    def get_existing_bucket(
+        self, project_id: Optional[int], minute: datetime
+    ) -> Optional[Bucket]:
+        for bucket in self.buckets:
+            if bucket.project_id == project_id and bucket.minute == minute:
+                return bucket
+        return None
+
+    def create_and_add_bucket(
+        self,
+        project_id: Optional[int],
+        start_minute: datetime,
+        processing_time: timedelta,
+    ) -> None:
+        bucket = self.get_existing_bucket(project_id, start_minute)
+        if not bucket:
+            bucket = Bucket(project_id, start_minute, processing_time)
+            self.buckets.append(bucket)
+        else:
+            bucket.add_processing_time_seconds(processing_time)
+
+    def write_to_bucket(
+        self, project_id: Optional[int], start: datetime, end: datetime
+    ) -> None:
+        start_minute = self.floor_minute(start)
+        left = start
+        right = self.ceil_minute(start)
+        while right <= end:
+            self.create_and_add_bucket(project_id, start_minute, right - left)
+            left = start_minute = right
+            right += timedelta(minutes=1)
+        self.create_and_add_bucket(project_id, start_minute, end - left)
+
+    def print_buckets(self) -> None:
+        for bucket in self.buckets:
+            print(bucket.minute, bucket.project_id, bucket.processing_time)
+
+    def compare_and_write_metric(self) -> None:
```

Comment:
This is another case where we are mixing the presentation with the  bucketing rate limiter logic. In order to keep the two responsibilities separate, please return the list of outstanding project and let the code that invoke this method decide what to do with that. 

Emitting the metrics inside ties the design of this class (which really is only about counting) with the purpose you are using this class for, which is instead relevant code in the replacer.
Example of how this approach breaks: Recording a metric in this case is only the initial step, we will want to replace this logic with something more drastic, which will demote the offending project ids to a separate queue.
When we will introduce that logic, the accounting logic will not change. The accounting policy stays the same, though we will have to touch this class because it combines the accounting function with the action to take. 
If you stick to one responsibility per class this will not happen, the counter will remain the same, which will be a sign of a better modularization.